lib/iris/util.py

```diff
@@ -2025,3 +2052,297 @@ def _strip_metadata_from_dims(cube, dims):
+            else:
+                raise NotImplementedError
+
+            data_masked = array_lib.ma.masked_array(
+                data, mask=mask, fill_value=self.uniform_fill_value
+            )
+
+            if self.is_cube(array_or_cube):
+                array_or_cube.data = data_masked
+            else:
+                array_or_cube = data_masked
+
+        return array_or_cube
+
+    def re_apply_masks(self, func_output):
+        """Apply mask to any returned un-masked arrays or Cubes."""
+        tuple_returned = isinstance(func_output, tuple)
+
+        if len(self.lift_addresses) > 0:
+            maskee = func_output
+            if not tuple_returned:
+                maskee = [func_output]
+
+            re_masked = [self.re_apply_mask(o) for o in maskee]
+            if tuple_returned:
+                result = tuple(re_masked)
+            else:
+                result = re_masked[0]
+
+        else:
+            result = func_output
+
+        return result
+
+    def restore_cube_masks(self):
+        """
+        Restore modified input cubes in-place, since we can't guarantee
+        they will all be within the result object.
+        """
+        for cube in self.cube_store:
+            self.re_apply_mask(cube)
+
+    @classmethod
+    def decorator(cls, decorated_func):
+        """
+        Temporarily convert input arrays/Cubes to use non-masked arrays if
+        all masks are ``False``.
+
+        Provided because working on non-masked
+        :obj:`~numpy.typing.ArrayLike`\\s
+        (e.g. :class:`numpy.ndarray` / :class:`dask.array.Array`)
+        can significantly improve performance. Any ``args`` or ``kwargs`` of
+        `decorated_func` that are :obj:`~numpy.typing.ArrayLike`\\s or
+        :class:`~iris.cube.Cube`\\s, and have all-``False`` masks,
+        are converted;
+        :class:`~iris.cube.Cube`\\s are modified in-place. After
+        `decorated_func`
+        has run, all returned :obj:`~numpy.typing.ArrayLike`\\s and
+        :class:`~iris.cube.Cube`\\s, together any modified input
+        :class:`~iris.cube.Cube`\\s (see above), are converted to use masked
+        arrays with the original ``False`` mask re-applied.
+
+        Parameters
+        ----------
+        decorated_func: callable
+            The callable object to be wrapped/decorated.
+
+        Returns
+        -------
+        Closure wrapped callable object.
+
+        Warns
+        -----
+        If inconsistent of False-mask-types or fill_values are present (see Notes).
+
+        Warnings
+        --------
+        Be careful what callables this is applied to. If any input objects
+        have an
+        all-``False`` mask, **all** returned non-masked objects will be
+        converted
+        to have an all-``False`` mask. And if the callable assigns
+        :obj:`~numpy.typing.ArrayLike`\\s/:class:`~iris.cube.Cube`\\s
+        in a way that can't be detected (e.g. modifying a global object, or
```

Comment:
It might be worth clarifying precisely what *can* be detected. I believe this is anything returned directly or returned as part of a tuple (correct me if I'm mistaken).

Reply:
6e0e7317e